infer/internal/ende/ende.go

```diff
@@ -293,16 +309,52 @@ func (e *ende) walkMap(
 	return resource.NewObjectProperty(result)
 }
 
+func prettyPrint(v interface{}) {
+	b, err := json.MarshalIndent(v, "", "  ")
+	if err == nil {
+		fmt.Println(string(b))
+	}
+}
+
 func (e *ende) Encode(src any) (resource.PropertyMap, mapper.MappingError) {
 	props, err := mapper.New(&mapper.Opts{
 		IgnoreMissing: true,
 	}).Encode(src)
 	if err != nil {
 		return nil, err
 	}
-	m := resource.NewObjectProperty(
-		resource.NewPropertyMapFromMap(props),
-	)
+
+	// If we see "asset" or "archive" properties who have the magic signatures inside, we assume we have an
+	// AssetOrArchive and need to pull the asset or archive out of the object.
+	// TODO,tkappler This is not safe: a user could have defined a property that looks like an
+	// AssetOrArchive but isn't.
+	m := resource.NewPropertyValueRepl(props,
+		nil, // keys are not changed
+		func(a any) (resource.PropertyValue, bool) {
+			if aMap, ok := a.(map[string]any); ok {
+				if rawAsset, ok := aMap["asset"]; ok {
```

Comment:
Instead of calling it `"asset"`, lets use a sig:
```suggestion
				if rawAsset, ok := aMap["somereallylongsigforassets"]; ok {
```

This way we can be pretty sure that if we see the key, it's what we are looking for.